### PR TITLE
An implementation of torch.tile as requested in pytorch/pytorch#38349

### PR DESCRIPTION
**Pull request by arbfay**
The approach is to simply reuse `torch.repeat` but adding one more functionality to tile, which is to prepend 1's to reps arrays if there are more dimensions to the tensors than the reps given in input. Thus for a tensor of shape (64, 3, 24, 24) and reps of (2, 2) will become (1, 1, 2, 2), which is what NumPy does.

I've encountered some instability with the test on my end, where I could get a random failure of the test (due to, sometimes, random value of `self.dim()`, and sometimes, segfaults). I'd appreciate any feedback on the test or an explanation for this instability so I can this.

### CODE REVIEW

**Comment by zou3519** [comment]
cc @mruberry for numpy compatibility - could you take a look at this please?

File: aten/src/ATen/native/TensorShape.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:848]: const auto here would save this line from casting on the rhs, I think (if there is to be a cast it should be a C++-style cast, not a C-style cast)

File: aten/src/ATen/native/TensorShape.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:851]: Egyptian-style braces for all if statements and for loops, even if their bodies are a single line

Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:851]: i here needs to be the same type as reps.size(), you can use decltype for this:

```
auto i = decltype(reps.size()){0}
```

File: test/test_torch.py
Comment (mruberry) [inline test/test_torch.py:2443]: We now require NumPy to run the test suite so this check isn't needed

File: test/test_torch.py
Comment (mruberry) [inline test/test_torch.py:2444]: This test should be device generic and placed in TestTorchDeviceType. See those tests for how they take a device argument.

File: test/test_torch.py
Comment (mruberry) [inline test/test_torch.py:2445]: For this test you'll want to enumerate a cross product of input shapes and reps values to test the fewer dims, more dims, and same dims cases. Also test for an empty input shape an an empty reps.

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:9843]: Lowercase tensor except when used as a Class specifier (like when doing "input (Tensor)" below)

File: torch/_tensor_docs.py
Comment (mruberry) [inline torch/_tensor_docs.py:3555]: I don't think you want the star on reps here?

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:9847]: "This function is similar to NumPy's tile function."

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:9850]: "A" -> "the"

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:9851]: "The" -> "the"

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:9853]: No need for a Returns: here. We typically only use it for complicated return values, like tuples of tensors.

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:9858]: Nice examples.

File: aten/src/ATen/native/TensorShape.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:856]: I think you need to use at::repeat here

**Review by mruberry — COMMENTED** [review]
Hi @arbfay, thanks for this contribution!

Overall this PR looks pretty good. I'm not totally sure why you're getting the error you are. It may be as simple as calling repeat instead of at::repeat. You might have to look at the intermediate values and debug the issue, thought.

I've left some comments that are mostly style/docs suggestions. The tests also need to be more thorough to account for more possible cases. 

Looking forward to reviewing the follow-up!

File: test/test_torch.py
Comment (mruberry) [reply test/test_torch.py:2445]: Also, do the test with both a tensor constructed in that shape and the transpose of the same tensor.

File: aten/src/ATen/native/TensorShape.cpp
Comment (arbfay) [reply aten/src/ATen/native/TensorShape.cpp:848]: Right, const and `static_cast`! 
I tend to use `auto` only when a type is cumbersome (containers within containers for example), otherwise, I stick to typing myself so it's not hidden from future devs. But I know some schools prefer using `auto` as much as possible. Is it the case for PyTorch?

File: aten/src/ATen/native/TensorShape.cpp
Comment (arbfay) [reply aten/src/ATen/native/TensorShape.cpp:851]: Didn't know about `decltype`, nice one!

File: torch/_tensor_docs.py
Comment (arbfay) [reply torch/_tensor_docs.py:3555]: well it works with *reps or a set (because reps is an IntArrayRef, somehow). I've seen other ops are using `*...` too so I just aligned, let me know if you think it's best not to

File: torch/_torch_docs.py
Comment (arbfay) [reply torch/_torch_docs.py:9858]: thanks!

File: torch/_torch_docs.py
Comment (arbfay) [reply torch/_torch_docs.py:9843]: What about simply adding "along each axis"?

File: aten/src/ATen/native/TensorShape.cpp
Comment (arbfay) [reply aten/src/ATen/native/TensorShape.cpp]: `at::native::repeat` instead

**Comment by arbfay** [comment]
Hi @mruberry,

Thanks for this review! I expected lots of comments for my first PR to PyTorch, and I got served 😄
For the failures, I've increased stability as much as I could with the last two commits. Hopefully, it will be enough now for all the pipelines. (Okay I will say it: it works perfectly in my environment)

Waiting now for your next review

File: aten/src/ATen/native/TensorShape.cpp
Comment (mruberry) [reply aten/src/ATen/native/TensorShape.cpp]: We allow a lot of latitude for individual style. As long as your types are correct you can use whatever you like.

**Comment by mruberry** [comment]
This failure on ROCm is real:

```
10:24:18 ======================================================================
10:24:18 ERROR [0.051s]: test_tile_cuda (__main__.TestTorchDeviceTypeCUDA)
10:24:18 ----------------------------------------------------------------------
10:24:18 Traceback (most recent call last):
10:24:18   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 834, in wrapper
10:24:18     method(*args, **kwargs)
10:24:18   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 272, in instantiated_test
10:24:18     result = test_fn(self, *args)
10:24:18   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 667, in only_fn
10:24:18     return fn(self, device, *args, **kwargs)
10:24:18   File "test_torch.py", line 6071, in test_tile
10:24:18     expected = np.tile(t.numpy(), dims)
10:24:18 TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.
10:24:18 
10:24:20 ----------------------------------------------------------------------
```

You can always use the pattern `t.cpu().numpy()`. This will convert CUDA tensors to CPU tensors (and then convert those to NumPy) and just directly convert CPU tensors to NumPy (the .cpu() will be a no-op for them).

Would you correct this test and I'll take a look after we get an updated test signal?

**Comment by arbfay** [comment]
@mruberry I think it's all ok with the tests (the one failing is because the pipeline is broken I think)

File: aten/src/ATen/native/TensorShape.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:854]: Use at::repeat instead of at::native::repeat here and below

File: test/test_torch.py
Comment (mruberry) [inline test/test_torch.py:6081]: No need to .cpu() or .numpy() the result. `assertEqual` will handle the needed conversions.

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:9844]: This looks pretty good. The last thing this needs to do is explain the padding behavior when there are more reps than dimensions in input, and more dimensions in input than values in reps. The NumPy documentation's description of these cases is actually pretty clear, so maybe that can be adapted?

**Review by mruberry — COMMENTED** [review]
Hey @arbfay!

Just a couple minor implementation issues and a small doc extension + test extension needed. Otherwise this PR looks good.

For the tests, this is currently missing autograd tests. That's understandable since we don't do a good job explaining how to add them.

You can automatically generate autograd tests by adding an entry to 

https://github.com/pytorch/pytorch/blob/68a3a3f3b55cbd18ccc580dfffc0f7c01fe4cda4/torch/testing/_internal/common_methods_invocations.py#L574

https://github.com/pytorch/pytorch/pull/46041 recently did this, too, so that might be a useful reference. Let me know if you have any questions.

File: aten/src/ATen/native/TensorShape.cpp
Comment (arbfay) [reply aten/src/ATen/native/TensorShape.cpp:854]: it doesn't compile with `at::repeat`. Otherwise, I can use `self.repeat(reps)`?

Comment (mruberry) [reply aten/src/ATen/native/TensorShape.cpp:854]: Sorry to mislead you. Using the method `self.repeat()` sounds great.

**Comment by arbfay** [comment]
Thanks @mruberry !

I've been struggling a bit with the autograd tests but it looks ok now. I've added more doc as asked, and used `self.repeat` instead of `at::native::repeat` for the return statements.

**Comment by mruberry** [comment]
I hope you don't mind but I tweaked the docs slight and tried to fix the common_methods() tests. If my fix for the tests didn't work properly then we can remove them for now to unblock this PR. Let's just wait for the CI to run.

**Comment by arbfay** [comment]
Thanks for your commits @mruberry, they are more than welcome! 

I've tried with your modification in common_methods and only one test is still a problem (no_reps_dims, I pushed a commit to remove it). This is great, I've been struggling to understand what was causing those failed assertions.

Let's wait for the pipelines.

**Comment by arbfay** [comment]
@mruberry is the PR good to go?

**Review by mruberry — APPROVED** [review]
Tests look good. Nice work, @arbfay!

**Comment by mruberry** [comment]
Congrats @arbfay! Thanks for implementing [torch.tile](https://pytorch.org/docs/master/generated/torch.tile.html?highlight=tile#torch.tile)!

**Comment by arbfay** [comment]
Thanks @mruberry, and thanks for the help! Been a nice ride, and learned a lot!

**Pull request closed without merging**
